src/DynamoCore/Search/NodeSearchModel.cs

```diff
@@ -65,5 +69,18 @@ private static void AddEntryToXml(XmlNode parent, NodeSearchElement entry)
             foreach (var entry in category.Entries)
                 AddEntryToXml(element, entry);
         }
+
+        public IEnumerable<LibraryItem> GetAllLibraryItemsByCategory(DynamoModel dynamoModel)
+        {
+            if (allLibraryItems == null || allLibraryItems.Count == 0)
+            {
+                allLibraryItems = new List<LibraryItem>();
+                foreach (var elem in SearchEntries)
+                {
+                    allLibraryItems.Add(new LibraryItem(elem, dynamoModel));
+                }
+            }
+            return allLibraryItems;
+        }
```

Comment:
There is a reason that `NodeSearchModel` does not contain `DynamoModel` -- because `NodeSearchModel` serves as just a *database* of nodes from which node instances can be created, it doesn't really care about which component owns it. In view of such design decision, it would be best if we don't construct and cache `allLibraryItems` here, it should all be done in `DynamoModel`. Here's what I think can happen:

1. Iterate through `dynamoModel.SearchModel.SearchEntries` to construct `allLibraryItems` from within `DynamoModel`, and then store it under `DynamoModel`.

2. What a `LibraryItem` needs is `Keyword` (search tags), `Parameters` and `ReturnKeys`, so don't pass `DynamoModel` to it, instead, construct a `NodeModel` to obtain `Keyword`, `Parameters` and `ReturnKeys`, before passing them to `LibraryItem` through its constructor (there's no need to store the `NodeModel`).

If you think that needs to be done as a separate PR, please create an internally tracked task and provide the link here. Otherwise, it can be done from this PR. Hope that helps.